### DISCUSSION TITLE
Saving/restoring window colors over remote control from a watcher

### DISCUSSION THREAD

**Discussion by rivenirvana** [kovidgoyal/kitty#7874] [Q&A] [answered]
I'm using this general watcher for some editor-specific kitty configs:
```py
from typing import Any, Dict

from kitty.boss import Boss
from kitty.window import Window


def on_cmd_startstop(boss: Boss, window: Window, data: Dict[str, Any]) -> None:
    command = data["cmdline"].strip().partition(" ")[0]
    programs = ("nv", "nvim", "vi", "vim", "nano")
    if command in programs:
        if data["is_start"]:
            boss.call_remote_control(window, ("disable_ligatures", "cursor"))
			# set predefined theme conf
            boss.call_remote_control(window, ("set-colors", "~/.config/kitty/editor-theme.conf"))
        else:
            boss.call_remote_control(window, ("disable_ligatures", "always"))
            # reset to configured theme on exit
            conf = boss.call_remote_control(window, ("get-colors", "-c"))
            args = ["=".join(line.split()) for line in conf.split("\n")]
            boss.call_remote_control(window, ("set-colors", *args))
```
And while this works, I was wondering if I can just easily use the escape codes over remote control as described [here](https://sw.kovidgoyal.net/kitty/color-stack/#saving-and-restoring-colors) to restore the colors on editor exit.
```
<ESC>]30001<ESC>\  # push to stack on program start
<ESC>]30101<ESC>\  # pop from stack on program exit
```
I've tried a couple of things, but I figured `@ launch --type=background` wouldn't work, and `@ run` didn't work either. I thought `@ send-text` would be the way to go, but no luck as well. Am I using `@ send-text` incorrectly?
```py
# does not work even without the printf
boss.call_remote_control(window, ("send-text", "printf '\x1b]30001\x1b\\'\r"))
boss.call_remote_control(window, ("send-text", "printf '\x1b]30101\x1b\\'\r"))
```
I checked boss.py too and there doesn't seem to be anything else I can use.

**Comment by rivenirvana** [comment]
Additionally:
- I've refrained from trying `boss.run_kitten()` to run the themes kitten non-interactively because it changes colors for the whole OS window and modifies `kitty.conf` and/or `current-theme.conf`.
- I've also refrained from just calling `@ set-colors` to `current-theme.conf`, since I like using the default theme and it has no key-values in its conf file to be read.

**Comment by kovidgoyal** [comment, marked as answer]
remote control doesnt send escape codes, if you want to send escape
codes you just write them to the tty device. How to do that depends on
your editor, there are some examples at
https://sw.kovidgoyal.net/kitty/mapping/#conditional-mappings-depending-on-the-state-of-the-focused-window
for vim and neovim. Which incidentally also means you dont need to try
to detect when an editor is running, instead the editor sets the user
variable and your watcher can react to that.

**Comment by rivenirvana** [comment]
For posterity's sake:
`.config/nvim/lua/autocommands.lua` # required in init.lua
```lua
vim.api.nvim_create_autocmd({ 'VimEnter', 'VimResume' }, {
  group = vim.api.nvim_create_augroup('KittySetVarVimEnter', { clear = true }),
  callback = function()
    io.stdout:write '\x1b]30001\x1b\\'
    io.stdout:write '\x1b]1337;SetUserVar=KITTY_IN_NVIM=MQ==\007'
  end,
})

vim.api.nvim_create_autocmd({ 'VimLeave', 'VimSuspend' }, {
  group = vim.api.nvim_create_augroup('KittyUnsetVarVimLeave', { clear = true }),
  callback = function()
    io.stdout:write '\x1b]30101\x1b\\'
    io.stdout:write '\x1b]1337;SetUserVar=KITTY_IN_NVIM\007'
  end,
})
```
`.config/kitty/scripts/editor-setup.py` # set as watcher in kitty.conf
```py
from typing import Any, Dict

from kitty.boss import Boss
from kitty.window import Window


def on_set_user_var(boss: Boss, window: Window, data: Dict[str, Any]) -> None:
    key, val = data["key"], data["value"]
    if key == "KITTY_IN_NVIM" and val is not None:
        boss.call_remote_control(window, ("disable_ligatures", "cursor"))
        boss.call_remote_control(
            window, ("set-colors", "~/.config/kitty/editor-theme.conf")
        )
    else:
        boss.call_remote_control(window, ("disable_ligatures", "always"))
```
